### ISSUE TITLE
Python 3.11 Windows ARM64

### ISSUE THREAD

**Issue by infinitepower18** [takluyver/pynsist#246]
Python has started offering embeddables and installers for Python 3.11 Windows ARM64. Is this something pynsist can include, if I want to distribute a native Windows ARM build?

https://www.python.org/downloads/release/python-3110a7/

You could probably add arm64 as an option for bitness in the config file.

**Comment by takluyver** [comment]
Downloading the embedded Python build would be easy enough. I don't know how to tell NSIS to make an ARM64 installer, though - can you find out about that?

I would also guess that few packages have pre-built Windows ARM64 wheels, which might limit how useful it is.

**Comment by infinitepower18** [comment]
The installer doesn't need to be in ARM64 I think. Right now there's not much info about Python on ARM64, since Python 3.11 is still in alpha stage. I guess once it becomes official we will see more packages getting ARM64 builds

**Comment by takluyver** [comment]
Ah, I see, NSIS will make an x86 exe which runs in an emulator on ARM Windows, according to https://sourceforge.net/p/nsis/feature-requests/547/ . At some point I guess it will be possible to build a native ARM64 installer, but not yet, it appears.

**Comment by infinitepower18** [comment]
Got a Raspberry Pi 400 which I can use to run Windows on ARM and the ARM version of Python 3.11 on, tested it with my program and it seems to be working perfectly fine without any changes.

![IMG_7368](https://user-images.githubusercontent.com/44692189/163998993-d8a710e8-a1ea-4291-a1e2-f52d4ae739b3.jpg)

So it's just a matter of pynsist adding the option for ARM64.

**Comment by takluyver** [comment]
Do you want to make a PR?

The code to make the URL to download Python from is here:

https://github.com/takluyver/pynsist/blob/0e05f3e6167bdbdb23e944bde7bb6c520cc0b9a5/nsist/__init__.py#L182-L190

In the config file, I'd suggest introducing a new parameter like `arch = arm64` as an alternative to the `bitness` option (though the old option needs to keep working). Here's the relevant code:

https://github.com/takluyver/pynsist/blob/0e05f3e6167bdbdb23e944bde7bb6c520cc0b9a5/nsist/configreader.py#L83

https://github.com/takluyver/pynsist/blob/0e05f3e6167bdbdb23e944bde7bb6c520cc0b9a5/nsist/configreader.py#L235

You'll also need to adapt anywhere else that uses `bitness`, like the `prepare_msvcrt` method, the `commands` module generating .exe wrappers, and the `wheels` module finding wheels to download.

**Comment by infinitepower18** [comment]
It would probably be better to wait until Python 3.11 goes out of beta. I somewhat got a messy implementation working, the installer will build and run fine on ARM systems via emulation and it will install the ARM64 version of the Python program. You can even build the installer on your regular x86 machine however for tkinter to work, you have to get the relevant files off an ARM Python installation and you need a device which can run Windows on ARM for that. So far I haven't tested builds with tkinter as I've now removed Windows from my Raspberry so have to go through the whole install process again.

I'll keep the changes to my own fork for now and will PR when the official release of 3.11 is out.

**Comment by takluyver** [comment]
Sounds good, thanks. :+1: 

Tkinter is kind of a pain in general - it's not in the embeddable builds, which is fair enough, but as far as I know there's no good way to get it without the full Python installer - there's no package on PyPI or anything. I'm not going to solve that, but if someone does, I hope Pynsist can take advantage of it.

**Comment by infinitepower18** [comment]
Happy to report that after building the ARM installer with the modified pynsist and including the ARM tkinter files my program successfully installs and runs in ARM64 native.

![Screenshot 2022-04-19 214035](https://user-images.githubusercontent.com/44692189/164092400-9858c659-64e4-4efc-bca2-2fbbc1b58d27.png)

However the installer is suggesting the Program Files (x86) directory by default when selecting Install for all users, so need to figure that out. There is a Program Files (Arm) folder but some sources are saying its for ARM32 apps. Notepad++ ARM64 version simply installs to the normal Program Files folder.

![Screenshot 2022-04-19 214121](https://user-images.githubusercontent.com/44692189/164092756-a5215a88-3e99-4662-8e45-98a3f5a7ad4e.png)

**Comment by takluyver** [comment]
There's a workaround in the code because the NSIS MultiUser extension used the wrong program files on x86-64 machines. Maybe that needs to be used on ARM too:

https://github.com/takluyver/pynsist/blob/0e05f3e6167bdbdb23e944bde7bb6c520cc0b9a5/nsist/pyapp.nsi#L258-L265

**Comment by infinitepower18** [comment]
Seems like it's not as easy as I thought.

While working on my next update of my program, I had to include one new package but turns out they don't have a specific arm64 Windows wheel yet. [Link](https://pypi.org/project/Pillow/#files)

The other packages I used if I'm not mistaken had [universal wheels](https://pypi.org/project/plyer/#files), which is why I didn't run into any problems.

So pynsist needs to look for specific Windows ARM64 wheels too.

**Comment by takluyver** [comment]
Yup, it's easy for pure Python projects, but quite probably a headache for any project with compiled extension modules.
